### DISCUSSION TITLE
about type system

### DISCUSSION THREAD

**Discussion by rraammiinn** [vlang/v#17402] [Questions and Answers]
can it be a gradual typed language ?

**Comment by bogen85** [comment]
That would require V to also be a dynamically typed language.
V is not dynamically typed, it is statically typed with type inference, and does have support for generics.

Adding dynamic typing to the core language would be a major change in direction and would fundamentally change the core language, and would not be the same language any more.

**Reply by rraammiinn** [reply]
I'm wondered if gradual typing is better than static typing for application development as static typing strictness isn't necessary for small apps developed by one person .

**Reply by JalonSolov** [reply]
Dynamic or even "gradual" typing is ok for prototyping code.  It does, however, open you to many possible problems which are simply impossible with statically typed code.

Anyone who has spent long debugging sessions only to find out the problem was that a variable wasn't the expected type will tell you that static typing is better.  With static typing, an int is always an int - it doesn't unexpectedly change to a string because of a function you called that you didn't realize was going to make the change.

With dynamic typing, that change can happen silently.  With static typing, the compiler won't let the code compile.

**Reply by bogen85** [reply]
Many developers developing small apps just for themselves want nothing to do with dynamic typing and all its pitfalls. Many accept the pitfalls because they like the dynamic typing.

However, gradual typing requires dynamic typing.
To introduce dynamic typing into an existing statically typed language is a major and extensive core change.

Since V started out as a statically typed language it is highly unlikely it would evolve to support dynamic typing. It already supports type inference and generics.

Gradual typing makes more sense for a language that starts out as dynamically typed. The other way around really does not make any sense from a design evolution (no matter how guided) standpoint. They are fundamentally different approaches to typing. Gradual just means the language supports both (at least to the programmer). Underneath the hood though that likely means it is a dynamically typed language with static typing extensions to have the appearance of being statically typed, simply because of how those extensions are enforced.

V is not dynamically typed, so gradual typing makes little sense.

**Reply by Wajinn** [reply]
> I'm wondered if gradual typing is better than static typing for application development as static typing strictness isn't necessary for small apps developed by one person .

I think dynamic and gradual typing makes some sense in the context of casual programmers making small scripts, because it's a bit easier to visually track and have everything in your head, and also as a matter of habit and style.  I think part of the problem is once a person is used to programming that way, it can be difficult to switch to a more strict and detailed style, and they may have initial resistance.  But on a professional or larger scale, dynamic becomes increasingly problematic and can even be dangerous.  Gradual is more like an intermediate step, which arguably has little place with V already being static and able to compile so quickly.

**Comment by bogen85** [comment]
When I say "fundamentally different" (dynamic and static, gradual implies both) I mean it is more than just a philosophical approach of the developers of V. They are radically different structural approaches. Adding dynamic typing (just to support gradual typing) would be a major rewrite. The result would no longer be V.

**Reply by rraammiinn** [reply]
I think v eventually will be an alternative to go and to some extend rust but not c and zig so it makes sense to be more concerned about application development rather than system programming and gradual typing guarantees high flexibility and a decent performance .

**Reply by JalonSolov** [reply]
Um... no, it doesn't.  Just search the web.  Here's a common theme...
```
The decision of using a dynamically-typed language versus a statically-typed one comes down to the core purpose
of a specific project -- namely, whether the goal is reliability or agility. Static typing is typically well
suited for building and maintaining programs that don't change that often, perform relatively straightforward
processes, and depend heavily on operational consistency. This method often works well in large-scale
application environments, where maintenance, reliability and security are priorities.

On the other hand, dynamically-typed languages are a better choice when the goal is to enable fast feature
additions, operational agility and overall application availability. Some also say dynamically-typed
languages are easier for novices to learn and write programs with, thanks to their non-prohibitive nature.
However, dynamic typing can quickly facilitate compounding coding mistakes and cause error-inducing bugs
to infiltrate application systems and perpetuate in production environments. Developers can effectively
offset the disadvantages of either by using things like a just-in-time compiler, but it's still a risky
choice when resiliency is a primary concern.
```
Basically, for systems programming (or anything else where you want stable, reliable code), you want static typing.  Personally, I'd prefer if dynamic typing didn't exist, since current static-typed languages (like Go and V) compile so fast that the main advantage of dynamic typing (speed of development) is no longer an issue.

Basically, you can type `v run my_program.v` and it will compile, link, and start running before the Python interpreter can even come up and start executing a script... that may be riddled with problems due to dynamic typing.

**Comment by rraammiinn** [comment]
I really like to languages like python take advantage of type hinting for increasing performance by they don't and I don't know why .
